[Merged by Bors] - feat(CategoryTheory): LCCC sections (constructing right adjoint to `Over.ChosenPullback.pullback`)

sinhp opened the pull request:
In LCCCs development, the `sections` functor is used to define the right adjoint to the pullback functor `Over.ChosenPullback.pullback`. 

This PR defines `Over.sections` and proves that it is a right adjoint to the `toOver (I : C) : C ⥤ Over I` (this is the computable analogue of `star`).

This is the computable enhancement of #22319: instead of `Limits.pullback` this PR builds on top of `Over.ChosenPullback` and thus we eliminate any instance of noncomputable constructs.

---
- [x] depends on: #31033
- [x] depends on: #31433

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

sinhp committed:
the preliminaries for locally cartesian closed categories (Sigma, Reindex, pull-push)

(Sigma=push, Reindex=pull, pull-push = cartesian product)

sinhp committed:
The section functor as a right adjoint to the star functor

sinhp committed:
The section functor as a right adjoint to the star functor

sinhp committed:
The section functor as a right adjoint to the star functor

sinhp committed:
fix

sinhp committed:
Merge remote-tracking branch 'refs/remotes/sina/lccc-prelim' into lccc-prelim

sinhp committed:
prelims for lcccs

sinhp committed:
fix for HasPullbacksAlong

sinhp committed:
parentheses and simp

sinhp committed:
Update Mathlib.lean

sinhp committed:
not needed assumption

sinhp committed:
rm simp

sinhp committed:
correcting attribution

sinhp committed:
Update Prelim.lean

sinhp committed:
Merge branch 'lccc-section' of https://github.com/sinhp/mathlib4 into lccc-section

sinhp committed:
ChosenPullback

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
pullbackObj

sinhp committed:
pullbackObj

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
ChosenPullback fst

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
ChosenPullback.hasPullbackAlong

sinhp committed:
ChosenPullback.pullback iso Over.pullback

sinhp committed:
remove second have

sinhp committed:
rm simp

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
rm have

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
naming

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
terminal simp

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
rm simp!

sinhp committed:
ChosenPullbacks C

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
HasPullbacks instance from ChosenPullbacks

sinhp committed:
Update ChosenPullback.lean

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
remove overMkHom

sinhp committed:
doc string

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
doc string added

sinhp committed:
Create types.lean

sinhp committed:
rm types from this

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
Create Types.lean

sinhp committed:
Delete Mathlib/CategoryTheory/LocallyCartesianClosed/Types.lean

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
pullbackIsoOverPullback_inv_app_

sinhp committed:
generalize pullbackMap

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
simp nf

sinhp committed:
Merge branch 'lccc-section' of https://github.com/sinhp/mathlib4 into lccc-section

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

sinhp committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into lccc-chosen_pullback

sinhp committed:
rm mapPullbackDesc

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
proof wo convert

sinhp committed:
Merge branch 'master' into lccc-chosen_pullback

sinhp committed:
doc string added

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
namespaces

sinhp committed:
Merge branch 'lccc-chosen_pullback' into lccc-section

sinhp committed:
computable star

sinhp committed:
Merge branch 'leanprover-community:master' into lccc-section

sinhp committed:
Delete Mathlib/CategoryTheory/LocallyCartesianClosed/Prelim.lean

sinhp committed:
Delete Mathlib/CategoryTheory/LocallyCartesianClosed/Sections0.lean

sinhp committed:
Update Sections.lean

sinhp committed:
Merge branch 'lccc-section' of https://github.com/sinhp/mathlib4 into lccc-section

sinhp committed:
Update Sections.lean

sinhp committed:
computable sections

sinhp committed:
use (Over.forget ..).mapIso

sinhp committed:
Merge branch 'leanprover-community:master' into lccc-section

sinhp committed:
toOver

sinhp committed:
linter

sinhp committed:
fix the damn linter; lake exe mk_all to get the order of names "right"

yikes

sinhp committed:
rm junk

sinhp committed:
ext instead of hom_ext

sinhp committed:
change fst to snd

this is because, as seen in lccc-section, we want to have
ChosenPullback.snd (toUnit X) (toUnit I) = CartesianMonoidalCategory.snd X I := rfl

sinhp committed:
doc string

sinhp committed:
Squashed version of lccc-chosen_pullback

sinhp committed:
Merge branch 'lccc-chosen_pullback-squashed' into lccc-section

sinhp committed:
not needed

sinhp committed:
Merge branch 'master' into lccc-section

sinhp committed:
Merge remote-tracking branch 'origin/master' into lccc-section

sinhp committed:
naming

sinhp committed:
naming

sinhp committed:
Merge branch 'lccc-chosen_pullback' into lccc-section

sinhp committed:
lemmas on interaction of toOver and ChosenPullback.pullback

sinhp committed:
toOverTerminal from toOverUnit

sinhp committed:
computable cartesian monoidal for slices

sinhp committed:
pullbackId and pullbackComp

sinhp committed:
rm junk

sinhp committed:
fixup! rm junk

sinhp committed:
Merge branch 'lccc-chosen_pullback' into lccc-section

sinhp committed:
make it computable

sinhp committed:
isoInv

sinhp committed:
Merge branch 'lccc-chosen_pullback' into lccc-section

sinhp committed:
variable scoping

sinhp committed:
add docstring

sinhp committed:
Update ChosenPullback.lean

sinhp committed:
Merge branch 'lccc-chosen_pullback' into lccc-section

sinhp committed:
rm unused import

sinhp committed:
cartesian monoidal section moved to a new PR

sinhp committed:
remove namespace Over

sinhp committed:
computable cartesian monoidal structure on slices

sinhp committed:
exe mk_all

sinhp committed:
rename to ChosenPullbackAlong

sinhp committed:
improve docstrings

sinhp committed:
Delete Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

sinhp committed:
add s

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-chosen_pullback

sinhp committed:
s

sinhp committed:
Merge branch 'lccc-chosen_pullback' of https://github.com/sinhp/mathlib4 into lccc-over

sinhp committed:
name change

sinhp committed:
Update Over.lean

joelriou commented on Mathlib/CategoryTheory/LocallyCartesianClosed/Sections.lean:
```diff
@@ -0,0 +1,259 @@
+
+variable {C : Type u₁} [Category.{v₁} C] [CartesianMonoidalCategory C]
+
+attribute [local instance] BraidedCategory.ofCartesianMonoidalCategory
```
My understanding is that if you need the braiding, the right assumptions are `[CartesianMonoidalCategory C] [BraidedCategory C]`.

sinhp replied:
You're likely correct, but let me try to understand what you are saying: Isn't every `CartesianMonoidalCategory` already braided? And `BraidedCategory.ofCartesianMonoidalCategory` proves it. 

If we add the variable `BraidedCategory C` instead, prima facie, it seems we prove a more general theorem, the braiding we use might be different then the particular one supplied by `BraidedCategory.ofCartesianMonoidalCategory`, but actually since we have an instance 
`Subsingleton (BraidedCategory C)` then it has be equal to `BraidedCategory.ofCartesianMonoidalCategory`. So, we don't actually get any more generality. I'm curious why we still would prefer `[BraidedCategory C]` to `BraidedCategory.ofCartesianMonoidalCategory`?

sinhp committed:
Merge branch 'master' of https://github.com/sinhp/mathlib4 into lccc-over

sinhp committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into lccc-over

sinhp committed:
sorry free

sinhp committed:
equivToOverUnit

sinhp committed:
Update Over.lean

robin-carlier commented on Mathlib/CategoryTheory/Comma/Over/Basic.lean:
```diff
@@ -326,6 +326,11 @@ def iteratedSliceBackward : Over f.left ⥤ Over f where
+@[simp]
+theorem iteratedSliceBackward_forget (f : Over X) :
+    iteratedSliceBackward f ⋙ Over.forget f = Over.map f.hom :=
+  rfl
```
I think we don’t want an equality of functors as a `[simp]` lemma (even when they are `rfl`): more often that not, this leads to this kind of lemmas firing in the type of expressions and leaving situations where we have to abuse some defeq to get out of.

Arguably, we should avoid recording equalities of functor to begin with (unless we really need to, e.g we intend to work in the 1-category of categories), and instead have something like 

```
@[simps!]
def iteratedSliceBackwardForgetIso (f : Over X) :
    iteratedSliceBackward f ⋙ Over.forget f ≅ Over.map f.hom :=
  Iso.refl _
```

which fits your use of this by feeding it to `eqToIso`.

But I see that a few equalities like this are recorded below (not as `[simp]` lemmas). So maybe it’s fine to keep this one and just remove the `[simp]`?

sinhp committed:
remove simp from equality of functors

https://github.com/leanprover-community/mathlib4/pull/30373#discussion_r2526336070

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
binaryFan_fst & binaryFan_snd

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Over.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
explicit vars

sinhp committed:
Merge branch 'lccc-over' of https://github.com/sinhp/mathlib4 into lccc-over

sinhp committed:
Merge branch 'master' of https://github.com/sinhp/mathlib4 into lccc-section

sinhp committed:
remove old merged file

sinhp committed:
Merge branch 'lccc-over' into lccc-section

sinhp committed:
fix after merge

sinhp committed:
abbrev

sinhp committed:
BraidedCategory C

sinhp committed:
Merge branch 'master' of https://github.com/sinhp/mathlib4 into lccc-over

sinhp committed:
rm mixin lemmas

dagurtomas committed:
Merge branch 'master' into lccc-over

dagurtomas committed:
Apply suggestion from @dagurtomas

dagurtomas committed:
move `@[expose] public section`

Expose the section for better accessibility.

sinhp committed:
Merge branch 'master' into lccc-over

sinhp committed:
Merge branch 'master' of https://github.com/sinhp/mathlib4 into lccc-over

sinhp committed:
Merge branch 'lccc-over' into lccc-section

sinhp committed:
Update Sections.lean

sinhp committed:
Update Mathlib.lean

sinhp committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib4 into lccc-section

sinhp committed:
@[expose] public section

sinhp committed:
remove Monoidal.Closed.Cartesian api

robin-carlier commented on Mathlib/CategoryTheory/LocallyCartesianClosed/Sections.lean:
```diff
@@ -0,0 +1,172 @@
+@[simps]
+def sections : Over I ⥤ C where
+  obj X := sectionsObj X
+  map u := sectionsMap u
```
(The github web interface is extremely about putting the diagram in backticks within the "suggestion" block, so I could not include it in the suggestion below, but I think it definitely should be in the docstring!)

robin-carlier commented on Mathlib/CategoryTheory/LocallyCartesianClosed/Sections.lean:
```diff
@@ -0,0 +1,172 @@
+
+variable [BraidedCategory C]
+
+/-- The currying operation `Hom ((star I).obj A) X → Hom A (I ⟹ X.left)`. -/
```
```suggestion
/-- The currying operation `Hom ((toOver I).obj A) X → Hom A (I ⟹ X.left)`. -/
```

robin-carlier commented on Mathlib/CategoryTheory/LocallyCartesianClosed/Sections.lean:
```diff
@@ -0,0 +1,172 @@
+    Equiv.symm_apply_apply] at w'
+  exact Over.homMk ((β_ A I).hom ≫ uncurry v₂) (by
+    dsimp [uncurry] at *
+    rw [Category.assoc, ← w', whiskerLeft_toUnit_comp_rightUnitor_hom, braiding_hom_fst])
```
For clarity, I think it’s better to keep all the "proof part" together, and not give the impression that this is "creating data with tactics"
```suggestion
  letI v₂ : A ⟶ (I ⟶[C] X.left) := v ≫ ChosenPullbacksAlong.fst (ihom I |>.map X.hom) (curryId I)
  Over.homMk ((β_ A I).hom ≫ uncurry v₂) (by
    have comm : toUnit A ≫ (curryId I) = v₂ ≫ (ihom I).map X.hom := by
      rw [IsTerminal.hom_ext isTerminalTensorUnit (toUnit A ) (v ≫ snd ..)]
      simp [v₂, condition]
    dsimp [curryId] at comm
    have w' := (ihom.adjunction I).homEquiv_naturality_right_square _ _ _ _ comm
    simp only [curriedTensor_obj_obj, curriedTensor_obj_map, curry,
      Equiv.symm_apply_apply] at w'
    dsimp [uncurry] at *
    rw [Category.assoc, ← w', whiskerLeft_toUnit_comp_rightUnitor_hom, braiding_hom_fst])
```

sinhp replied:
other than trying to understand this, I think your suggestion makes sense, so I'll commit it. thanks.

robin-carlier commented on Mathlib/CategoryTheory/LocallyCartesianClosed/Sections.lean:
```diff
@@ -0,0 +1,172 @@
+    invFun := sectionsUncurry
+    left_inv {u} := sections_uncurry_curry
+    right_inv {v} := sections_curry_uncurry
+  }
```
```suggestion
  homEquiv A X :=
    { toFun := sectionsCurry
      invFun := sectionsUncurry
      left_inv _ := sections_uncurry_curry
      right_inv _ := sections_curry_uncurry }
```

robin-carlier commented on Mathlib/CategoryTheory/LocallyCartesianClosed/Sections.lean:
```diff
@@ -0,0 +1,172 @@
+    · aesop
+
+variable (I) in
+/-- The adjunction between the star functor and the sections functor. -/
```
```suggestion
/-- The adjunction between the toOver functor and the sections functor. -/
```

robin-carlier commented on Mathlib/CategoryTheory/LocallyCartesianClosed/Sections.lean:
```diff
@@ -0,0 +1,172 @@
+
+/-- An auxiliary definition which is used to define the adjunction between the star functor
+and the sections functor. See starSectionsAdjunction`. -/
+def coreHomEquivToOverSections : CoreHomEquiv (toOver I) (sections I) where
```
```suggestion
@[simps homEquiv]
def coreHomEquivToOverSections : CoreHomEquiv (toOver I) (sections I) where
```

This annotation is needed so that the simps lemmas for `toOverSectionsAdj` refer to `sectionCurry` and `sectionUncurry` instead of this def.

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Sections.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Sections.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Sections.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Sections.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Sections.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Sections.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Sections.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
name change

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Sections.lean

Co-authored-by: Robin Carlier <57142648+robin-carlier@users.noreply.github.com>

sinhp committed:
fix names

sinhp committed:
better variable scoping and style

sinhp committed:
better docstring

robin-carlier reviewed (commented):
Thanks!

maintainer merge

jcommelin commented on Mathlib/CategoryTheory/LocallyCartesianClosed/Sections.lean:
```diff
@@ -0,0 +1,153 @@
+    rw [Category.assoc, ← w', whiskerLeft_toUnit_comp_rightUnitor_hom, braiding_hom_fst])
+
+@[simp]
+theorem sections_curry_uncurry {X : Over I} {A : C} {v : A ⟶ (sections I).obj X} :
```
```suggestion
theorem sectionsCurry_sectionUncurry {X : Over I} {A : C} {v : A ⟶ (sections I).obj X} :
```

jcommelin reviewed (commented):
Thanks :tada:

bors d+

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Sections.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

sinhp committed:
Update Mathlib/CategoryTheory/LocallyCartesianClosed/Sections.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

sinhp commented:
> ✌️ sinhp can now approve this pull request. To approve and merge a pull request, simply reply with `bors r+`. More detailed instructions are available [here](https://bors.tech/documentation/getting-started/#reviewing-pull-requests).

bors r+

sinhp committed:
fix

sinhp committed:
Merge branch 'master' into lccc-section
